### ISSUE TITLE
Reduce Action Space from 19 to 7 Actions

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#53]
## Problem

**19-action space is too large** for PPO to explore effectively with standard hyperparameters.

**Evidence:**
- Required exploration: ~19,000 steps
- Actual exploration: ~2,000-5,000 steps
- **DEFICIT: 10x insufficient**
- Agent finds Action 6 early, never explores 13+ other actions

---

## Proposed Solution: Reduce Action Space

**Simplify from 19 to 7 actions** by keeping only the most promising SL/TP combinations.

### Current Action Space (19 actions)

```
Action 0: Hold
Actions 1-9: Long positions
  1: Long (SL=2%, TP=4%)
  2: Long (SL=2%, TP=6%)
  3: Long (SL=2%, TP=8%)
  4: Long (SL=5%, TP=4%)
  5: Long (SL=5%, TP=6%)
  6: Long (SL=5%, TP=8%)  ← Agent's favorite
  7: Long (SL=10%, TP=4%)
  8: Long (SL=10%, TP=6%)
  9: Long (SL=10%, TP=8%)

Actions 10-18: Short positions (mirror of long)
```

**Problems:**
- Too many similar actions (e.g., SL=5% TP=4% vs SL=5% TP=6%)
- Redundant coverage of risk/reward space
- Agent can't explore thoroughly

### Proposed Action Space (7 actions)

Keep only **diverse, well-spaced risk/reward profiles**:

```
Action 0: Hold
Action 1: Long Conservative (SL=2%, TP=4%)   - R:R = 2.0
Action 2: Long Moderate (SL=5%, TP=8%)       - R:R = 1.6
Action 3: Long Aggressive (SL=10%, TP=15%)   - R:R = 1.5
Action 4: Short Conservative (SL=2%, TP=4%)  - R:R = 2.0
Action 5: Short Moderate (SL=5%, TP=8%)      - R:R = 1.6
Action 6: Short Aggressive (SL=10%, TP=15%)  - R:R = 1.5
```

**Benefits:**
- **3x fewer actions** = 3x more samples per action
- Clear risk stratification (conservative/moderate/aggressive)
- Still covers full strategy space (long/short, tight/wide stops)
- Forces agent to learn shorts (can't ignore with only 3 long actions)

---

### Option A: New Environment Config

```python
# In FuturesOneStepEnvConfig
@dataclass
class FuturesOneStepEnvConfig:
    # Simple action space (7 actions)
    stoploss_levels: List[float] = field(default_factory=lambda: [-0.02, -0.05, -0.10])
    takeprofit_levels: List[float] = field(default_factory=lambda: [0.04, 0.08, 0.15])
    
    # Note: Creates 1 hold + 3 long + 3 short = 7 actions
    # vs old: 1 hold + 9 long + 9 short = 19 actions
```

### Option B: Action Selection Strategy

Instead of Cartesian product of all SL/TP combinations, select specific pairs:

```python
def create_simplified_action_map() -> Dict[int, tuple]:
    """Create 7-action space with specific SL/TP pairs."""
    action_map = {
        0: (None, None, None),                    # Hold
        1: ("long", -0.02, 0.04),                # Conservative
        2: ("long", -0.05, 0.08),                # Moderate
        3: ("long", -0.10, 0.15),                # Aggressive
        4: ("short", -0.02, 0.04),               # Conservative
        5: ("short", -0.05, 0.08),               # Moderate
        6: ("short", -0.10, 0.15),               # Aggressive
    }
    return action_map
```

### Update Environment

```python
class FuturesOneStepEnvSimple(FuturesOneStepEnv):
    """Simplified 7-action version of FuturesOneStepEnv."""
    
    def __init__(self, df, config, feature_preprocessing_fn=None):
        # Override action map creation
        self.action_map = create_simplified_action_map()
        self.action_spec = Categorical(7)  # 7 actions instead of 19
        
        # Rest of initialization same as parent
        super().__init__(df, config, feature_preprocessing_fn)
```

---

### Exploration Improvement

```
19-action space:
  Exploration budget: 2,000-5,000 steps
  Per-action samples: ~100-250
  Result: Insufficient, single-action dominance

7-action space:
  Exploration budget: 2,000-5,000 steps
  Per-action samples: ~300-700
  Result: 3x more data per action = better learning
```

### Success Metrics

| Metric | 19-Action Baseline | 7-Action Target |
|--------|-------------------|-----------------|
| Actions with >5% usage | 1-2 | > 5 |
| Action entropy | 0.48 | > 1.2 |
| Single action dominance | 87% | < 30% |
| Short position usage | 0-5% | > 25% |
| Test Sharpe ratio | -5.77 | > -2.0 |

---

### Why These SL/TP Pairs?

**Conservative (SL=2%, TP=4%)**
- Tight stop-loss limits losses
- 2:1 R:R ratio (good risk management)
- Suitable for ranging/choppy markets

**Moderate (SL=5%, TP=8%)**
- Balanced risk/reward
- 1.6:1 R:R ratio
- Agent's historical favorite (Action 6 in 19-space)
- Suitable for moderate trends

**Aggressive (SL=10%, TP=15%)**
- Wide stop allows price fluctuation
- 1.5:1 R:R ratio
- Captures large moves
- Suitable for strong trends

### Why Not More Granular?

**Problem with fine-grained SL/TP:**
- SL=5% TP=6% vs SL=5% TP=8% → negligible difference in practice
- Agent wastes exploration samples on near-identical strategies
- Adds complexity without strategic value

---

### Week 1: Environment Update
- [ ] Create `FuturesOneStepEnvSimple` class or config flag
- [ ] Update action map to 7 actions
- [ ] Update action spec
- [ ] Test environment (rollouts, resets)

### Week 2: Training Experiments
- [ ] Baseline: 19-action space (reproduce previous results)
- [ ] Test: 7-action space (same hyperparameters)
- [ ] Compare action distributions
- [ ] Compare learning curves
- [ ] Measure time-to-convergence

### Week 3: Analysis & Optimization
- [ ] Analyze which actions learned best
- [ ] Try alternative SL/TP selections
- [ ] Test with increased entropy coefficient (#50)
- [ ] Test with curriculum learning (#49)

---

## Configuration Example

```yaml
# config.yaml
env:
  name: FuturesOneStepEnv
  action_space_type: "simple"  # or "full"
  
  # Simple action space (7 actions)
  stoploss_levels: [-0.02, -0.05, -0.10]
  takeprofit_levels: [0.04, 0.08, 0.15]
  
  # Full action space would be:
  # stoploss_levels: [-0.02, -0.05, -0.10]
  # takeprofit_levels: [0.04, 0.06, 0.08]
  # (Creates 1 + 3×3 + 3×3 = 19 actions)
```

---

## Advantages vs Other Approaches

**vs Curiosity (#48):**
- ✅ Simpler (no new modules)
- ✅ Directly addresses root cause (too many actions)
- ✅ Faster training (fewer actions to explore)

**vs Curriculum (#49):**
- ✅ No phased training complexity
- ✅ Simpler implementation
- ❌ Less structured exploration

**vs Entropy Scheduling (#50):**
- ✅ Addresses problem at source (action space size)
- ✅ Can be combined for even better results

**vs Hierarchical (#51):**
- ✅ Much simpler architecture
- ❌ Less adaptable to market regimes

**Best approach:** Try 7-action space **first** (simplest), then add other techniques if needed.

---

### Q: Will 7 actions be sufficient to capture trading strategies?

**A:** Yes. The 7 actions cover:
- 3 risk levels (conservative, moderate, aggressive)
- 2 directions (long, short)
- 1 hold option

This is similar to what human traders use (they don't micro-optimize every 1% of SL/TP).

### Q: What if optimal SL/TP is between our selected levels?

**A:** 
1. Agent can still profit with nearby levels
2. If needed, add 1-2 intermediate actions (still << 19)
3. Fine-grained optimization can come later via meta-learning

### Q: Won't this reduce model capacity?

**A:** No. The agent still has full policy network capacity. We're reducing **action space complexity**, not **model capacity**. This actually helps learning by providing more data per action.

---

### Option B: R:R-Based Selection
```python
actions = [
    (None, None, None),           # Hold
    ("long", -0.03, 0.06),        # R:R = 2.0
    ("long", -0.05, 0.08),        # R:R = 1.6
    ("long", -0.08, 0.08),        # R:R = 1.0
    ("short", -0.03, 0.06),       # R:R = 2.0
    ("short", -0.05, 0.08),       # R:R = 1.6
    ("short", -0.08, 0.08),       # R:R = 1.0
]
```

### Option C: Market-Regime Optimized
```python
# Based on volatility analysis of BTC
actions = [
    (None, None, None),           # Hold
    ("long", -0.02, 0.04),        # Low volatility
    ("long", -0.06, 0.10),        # Medium volatility
    ("long", -0.12, 0.20),        # High volatility
    ("short", -0.02, 0.04),       # Low volatility
    ("short", -0.06, 0.10),       # Medium volatility
    ("short", -0.12, 0.20),       # High volatility
]
```

---

## References

**Action Space Design:**
- Dulac-Arnold et al. (2015) - "Deep Reinforcement Learning in Large Discrete Action Spaces"
- Chandak et al. (2019) - "Learning Action Representations for Reinforcement Learning"

**Trading-Specific:**
- Most professional trading systems use 3-5 risk levels, not 100s of micro-variations

---

## Related Issues

- #48 - Curiosity (easier with 7 actions than 19)
- #49 - Curriculum (7-action base can expand to 13, then 19)
- #50 - Entropy scheduling (more effective with smaller action space)
- #51 - Hierarchical (can use 7-action sub-policies)
- #52 - PBT (faster with 7 actions)
